python/tests/test_edges.py

```diff
@@ -190,5 +190,75 @@ def on_complete():
     assert on_next_count == 4
 
 
+@dataclasses.dataclass
+class MyCustomClass:
+    value: int
+    name: str
+
+
+def test_multi_segment():
+    def segment_source(seg: srf.Builder):
+        # Use a generator function as the source
+        def source_gen():
+            print("Source: Starting")
+            for i in range(5):
+                yield MyCustomClass(i, "Instance-{}".format(i))
+                # yield m.DerivedA()
+
+            print("Source: Complete")
+
+        # Create the source object
+        # source = seg.make_source("source", source_gen)
+        source = m.SourceDerivedB(seg, "source")
+        source.launch_options.pe_count = 1
+
+        egress = seg.get_egress("port1")
+
+        seg.make_edge(source, egress)
+
+    def segment_sink(seg: srf.Builder):
+        ingress = seg.get_ingress("port1")
+
+        # This method will get called each time the sink gets a value
+        def sink_on_next(x: MyCustomClass):
+            print("Sink: Got Obj Name: {}, Value: {}".format(x.name, x.value))
+
+        # Build the sink object
+        # sink = seg.make_sink("sink", sink_on_next, None, None)
+        sink = m.SinkBase(seg, "sink")
+
+        seg.make_edge(ingress, sink)
```

Comment:
If you replace `MyCustomClass` with a C++ wrapped type, `m.DerivedA`, like you have in the comment, we will get into a situation that I think we want to avoid if we can. Using C++ wrapped types here will cause an unnecessary amount of casing back and forth between python and C++. In a single segment scenario this would go C++ -> C++ and never invoke the GIL. But with multiple segments and ports its going to go C++ -> Python -> Python -> C++ even though we have C++ types and C++ nodes and have the converters registered.

To avoid this, I think we will need some ability to specify the port type from the python side. Something like:
```python
pipeline.make_segment("segment_source", [], [(m.DerivedA, "port1")], segment_source)

pipeline.make_segment("segment_sink", [(m.DerivedA, "port1")], [], segment_sink)
```
where instead of just providing the port name, you provide a tuple of `(port_type, port_name)`. This would allow the runtime the ability to look up the type in pybind11 and see if there was a C++ registered class. If so, we could build the ports in a way that doesnt require the python cast.

Reply:
Resolved by moving adapter's to be runtime constructs.